File: heaps/lists_merge/lists_merge.cpp
#include <vector>
#include <queue>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode *mergeSortedLists(vector<ListNode *> &lists)
{ // LeetCode Q.23.
    // Min heap. Format: {value, its linked list idx}.
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;

    vector<ListNode *> listsNodes;
    for (int idx = 0; idx < lists.size(); idx++)
    {
        if (lists[idx] != nullptr)
            minHeap.push({lists[idx]->val, idx});

        listsNodes.push_back(lists[idx]);
    }

    vector<int> sortedValues;

    while (!minHeap.empty())
    {
        auto [value, listIdx] = minHeap.top();
        sortedValues.push_back(value);
        minHeap.pop();

        if (listsNodes[listIdx]->next != nullptr)
        {
            listsNodes[listIdx] = listsNodes[listIdx]->next;
            minHeap.push({listsNodes[listIdx]->val, listIdx});
        }
    }

    if (sortedValues.empty())
        return nullptr;

    ListNode *sortedList = new ListNode(sortedValues[0]);
    ListNode *currentNode = sortedList;

    for (size_t idx = 1; idx < sortedValues.size(); idx++)
    {
        currentNode->next = new ListNode(sortedValues[idx]);
        currentNode = currentNode->next;
    }

    return sortedList;
}

File: heaps/lists_merge/lists_merge.py
import heapq


class ListNode:
    def __init__(self, val=0, next_node=None) -> None:
        self.val = val
        self.next = next_node


def merge_sorted_lists(lists: list[ListNode | None]) -> ListNode | None:  # LeetCode Q.23.
    min_heap = []  # Format: (value, its linked list idx).

    lists_nodes = []
    for idx, node in enumerate(lists):
        if node is not None:
            heapq.heappush(min_heap, (node.val, idx))
        lists_nodes.append(node)

    sorted_list, current_node = None, None
    while min_heap:
        value, list_idx = heapq.heappop(min_heap)
        if sorted_list is None:
            sorted_list = ListNode(value)
            current_node = sorted_list

        else:
            current_node.next = ListNode(value)
            current_node = current_node.next

        lists_nodes[list_idx] = lists_nodes[list_idx].next  # Next node in this list.
        if lists_nodes[list_idx] is not None:
            heapq.heappush(min_heap, (lists_nodes[list_idx].val, list_idx))

    return sorted_list
